crates/context/config/src/client/protocol/icp.rs

```diff
@@ -0,0 +1,262 @@
+    #[error("updating canister")]
+    Mutate,
+}
+
+impl ProtocolTransport for IcpTransport<'_> {
+    type Error = IcpError;
+
+    async fn send(
+        &self,
+        request: TransportRequest<'_>,
+        payload: Vec<u8>,
+    ) -> Result<Vec<u8>, Self::Error> {
+        let Some(network) = self.networks.get(&request.network_id) else {
+            return Err(IcpError::UnknownNetwork(request.network_id.into_owned()));
+        };
+
+        let canister_id = Principal::from_text(&request.contract_id)
+            .map_err(|_| IcpError::InvalidCanisterId(request.contract_id.into_owned()))?;
+
+        match request.operation {
+            Operation::Read { method } => {
+                network
+                    .query(&canister_id, method.into_owned(), payload)
+                    .await
+            }
+            Operation::Write { method } => {
+                network
+                    .mutate(&canister_id, method.into_owned(), payload)
+                    .await
+            }
+        }
+    }
+}
+
+impl Network {
+    async fn query(
+        &self,
+        canister_id: &Principal,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, IcpError> {
+        self.client
+            .fetch_root_key()
+            .await
+            .map_err(|_| IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Failed to fetch root key".to_string(),
+            })?;
+
+        let response = self
+            .client
+            .query(canister_id, method)
+            .with_arg(args)
+            .call()
+            .await;
+
+        response.map_or(
+            Err(IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Error while quering".to_string(),
+            }),
+            |response| Ok(response),
+        )
+    }
+
+    async fn mutate(
+        &self,
+        canister_id: &Principal,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, IcpError> {
+        self.client
+            .fetch_root_key()
+            .await
+            .map_err(|_| IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Failed to fetch root key".to_string(),
+            })?;
+
+        let response = self
+            .client
+            .update(canister_id, method)
+            .with_arg(args)
+            .call()
+            .await;
+
+        match response {
+            Ok(CallResponse::Response((data, _))) => {
+                // Now try to decode the actual data
+                match candid::decode_one::<Result<(), String>>(&data) {
+                    Ok(decoded) => {
+                        match decoded {
+                            Ok(()) => Ok(vec![]),                     // Return empty vec for success
+                            Err(err_msg) => Ok(err_msg.into_bytes()), // Return error message as bytes
+                        }
+                    }
```

Comment:
this only encodes errors as the response? how about the actual return data, can we make that into `Vec<u8>`, without the type information that ICP typically infuses?